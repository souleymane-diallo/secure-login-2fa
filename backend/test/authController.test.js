const request = require('supertest');
const { describe, it, expect } = require('@jest/globals');
const app = require('../app');

let server;

beforeAll((done) => {
  const PORT = 5001; 
  server = app.listen(PORT, () => {
    console.log(`Test server running on port ${PORT}`);
    done();
  });
});

afterAll((done) => {
  server.close(done);
});

describe('Auth Controller', () => {
  describe('POST /api/register', () => {
    it('should register a user with valid email and password', async () => {
      const validPayload = {
        email: `newuser${Date.now()}@example.com`,  // Utiliser un email unique pour chaque test
        password: 'password123'
      };
      const response = await request(app)
        .post('/api/register')
        .send(validPayload);

      expect(response.status).toBe(201);
      expect(response.body).toHaveProperty('message', 'User registered successfully');
    });

    it('should reject short passwords during registration', async () => {
      const shortPasswordPayload = {
        email: `test${Date.now()}@example.com`,
        password: 'short'
      };
      const response = await request(app)
        .post('/api/register')
        .send(shortPasswordPayload);

      expect(response.status).toBe(400);
      expect(response.body).toHaveProperty('error', 'Password must be at least 6 characters long.');
    });

    it('should not allow SQL injection through email during registration', async () => {
      const maliciousPayload = {
        email: "' OR 1=1 --",
        password: 'password123'
      };
      const response = await request(app)
        .post('/api/register')
        .send(maliciousPayload);

      expect(response.status).toBe(400);
      expect(response.body).toHaveProperty('error');
    });
  });

  describe('POST /api/login', () => {
    it('should not allow SQL injection through email', async () => {
      const maliciousPayload = {
        email: "' OR 1=1 --",
        password: 'password123'
      };
      const response = await request(app)
        .post('/api/login')
        .send(maliciousPayload);

      expect(response.status).toBe(400);
      expect(response.body).toHaveProperty('error');
    });

    it('should not allow SQL injection through password', async () => {
      const maliciousPayload = {
        email: 'test@example.com',
        password: "' OR 1=1 --"
      };
      const response = await request(app)
        .post('/api/login')
        .send(maliciousPayload);

      expect(response.status).toBe(400);
      expect(response.body).toHaveProperty('error');
    });

    it('should return 2FA setup required for new users', async () => {
      const newUserPayload = {
        email: `newuser${Date.now()}@example.com`,
        password: 'password123'
      };

      // Register the new user first
      await request(app)
        .post('/api/register')
        .send(newUserPayload);

      // Attempt to login
      const response = await request(app)
        .post('/api/login')
        .send(newUserPayload);

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty('message', '2FA setup required');
      expect(response.body).toHaveProperty('qrCodeUrl');
      expect(response.body).toMatchSnapshot();
    });
  });
});
