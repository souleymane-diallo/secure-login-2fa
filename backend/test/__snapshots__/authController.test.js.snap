// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Controller POST /api/login should return 2FA setup required for new users 1`] = `
{
  "message": "2FA setup required",
  "qrCodeUrl": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAALQAAAC0CAYAAAA9zQYyAAAAAklEQVR4AewaftIAAAdsSURBVO3BQY4kRxLAQDLQ//8yd45+SiBR1SMp1s3sD9a6xGGtixzWushhrYsc1rrIYa2LHNa6yGGtixzWushhrYsc1rrIYa2LHNa6yGGtixzWushhrYv88CGVv6liUnmjYlL5J1VMKp+o+CaVv6niE4e1LnJY6yKHtS7yw5dVfJPKGxVPVJ5UTCpTxaQyVbyh8omKSWWqmFSmijcqvknlmw5rXeSw1kUOa13kh1+m8kbFJ1SeVDxRmSqeVDxR+UTFpPKGyhOVqeINlTcqftNhrYsc1rrIYa2L/PAfV/FE5RMq31QxqbxR8UbFE5WbHNa6yGGtixzWusgP/3EqU8VU8U0VT1SeqEwVk8qkMlU8qXiicrPDWhc5rHWRw1oX+eGXVfymikllqniiMlVMKm9UfKLiicpU8URlqvimin+Tw1oXOax1kcNaF/nhy1T+JpWpYlKZKr5JZaqYVKaKSWWqmFSmikllqnhDZap4ovJvdljrIoe1LnJY6yL2B/9HVJ5UvKHymyo+ofJGxX/ZYa2LHNa6yGGti/zwIZWp4onK31QxVUwqT1TeqHhDZaqYVKaKSeVJxRsqU8UTlaliUnmj4hOHtS5yWOsih7Uu8sOHKiaVqeITFU9UpoonKlPFpPKk4g2Vb1KZKiaVSWWq+ITKVPGJim86rHWRw1oXOax1EfuDL1J5UjGpTBWTylTxRGWqeKLyRsUTlaliUpkqJpUnFZPKVDGpTBWTylTxhspU8YbKVPGJw1oXOax1kcNaF7E/+IDKVDGpPKl4Q+VJxaQyVbyh8qRiUvlExTepTBWfUJkq3lB5UvGJw1oXOax1kcNaF/nhy1Smiicqb1Q8UfmEypOKSeVvUnmjYlL5m1Smiknlmw5rXeSw1kUOa13E/uAXqUwVn1D5popJZar4TSqfqJhU3qh4Q+VJxROVqeKbDmtd5LDWRQ5rXeSHD6lMFU9UPlExqTypmFQmlaliUpkq/qaKJyqfUHlS8YbKk4rfdFjrIoe1LnJY6yI/fJnKGxWTylTxRsUnVKaKJypTxaTyRsUTlaniicpUMak8UZkqnqg8UXlS8YnDWhc5rHWRw1oX+eHLKp6oPKmYVKaKqWJSmSqmiknlDZWpYlL5hMpUMVVMKlPFE5UnFf9lh7UucljrIoe1LmJ/8EUqTyq+SeWNiicqU8UTlaniicpU8URlqphUpoo3VN6omFTeqPhNh7UucljrIoe1LvLDh1R+k8qTijdUpoqpYlJ5Q+VJxaQyVTxRmSomlScVU8UbKt+kMlV84rDWRQ5rXeSw1kV++GUVk8pU8aTiicqTiqniicqTiicq/yYVb6hMFVPFpDJVTCqTylTxTYe1LnJY6yKHtS5if/ABlScVn1CZKp6oPKmYVL6pYlJ5UvEJlaliUpkqnqg8qZhUpoonKlPFNx3WushhrYsc1rrID19W8U0Vk8obFU8qJpWpYlKZKp5UTCqTyicqJpWpYlKZKp5UfEJlqphUpopPHNa6yGGtixzWusgP/zCVNyomlaniN1VMKlPFJyqeqEwq31TxTRVPKr7psNZFDmtd5LDWRewPvkhlqnii8qRiUpkqnqhMFZPKk4o3VJ5UTCp/U8UbKlPFpPJGxaQyVXzisNZFDmtd5LDWRewP/iKVqWJS+aaKSeWNiknlb6p4ojJVPFF5o+I3qUwVnzisdZHDWhc5rHUR+4MPqEwVk8pUMalMFU9U3qiYVKaKSeVJxaQyVbyhMlU8UXmj4onKJyqeqEwVv+mw1kUOa13ksNZFfvgylaliUnlDZaqYVKaKSeWNiknlEypvqEwVv6liUnlS8QmVJxWfOKx1kcNaFzmsdRH7gw+oTBWTyicqJpWpYlKZKiaVqeINlaniicqTit+kMlW8ofKk4g2VJxWfOKx1kcNaFzmsdZEfflnFJ1SeqHxCZap4Q+VJxSdUpopJ5UnFE5VPqDypmCp+02GtixzWushhrYvYH/yHqUwVk8pU8U0qU8UTlaniEypTxaTyRsUbKm9U/KbDWhc5rHWRw1oX+eFDKn9TxRsVT1Smikllqpgqnqg8UXlSMam8UfFE5YnKVPGkYlKZVKaKbzqsdZHDWhc5rHWRH76s4ptU3lB5UjFVPKn4myomlaliUplUvqniDZWpYlKZVKaKTxzWushhrYsc1rrID79M5Y2KT1S8ofJPqphUpoo3KiaVN1Q+UfFPOqx1kcNaFzmsdZEfLqPyRsWk8omKJyqTylQxqUwVb1RMKlPFpPKk4onKVPE3Hda6yGGtixzWusgP/3EVT1SmiknljYonKlPFVDGpPKn4hMpUMak8qZhUnlT8kw5rXeSw1kUOa13kh19W8W9W8YbKJ1TeUHlS8aTijYpJZar4NzusdZHDWhc5rHWRH75M5W9SeVIxqUwVT1SmiknlicpU8UTlScWkMlVMKk8qPqEyVTypmFS+6bDWRQ5rXeSw1kXsD9a6xGGtixzWushhrYsc1rrIYa2LHNa6yGGtixzWushhrYsc1rrIYa2LHNa6yGGtixzWushhrYv8D2kl2m5amvHoAAAAAElFTkSuQmCC",
  "user": {
    "email": "newuser1716735782622@example.com",
    "id": 10,
  },
}
`;
